Augment a WTF::URL helper function to return a list of query parameters

whsieh opened the pull request:
#### 71a2925a0afe123fc95af80b8336eeb93813fe9d
<pre>
Augment a WTF::URL helper function to return a list of query parameters
<a href="https://bugs.webkit.org/show_bug.cgi?id=250402">https://bugs.webkit.org/show_bug.cgi?id=250402</a>

Reviewed by Chris Dumez.

Tweak `removeQueryParameters` to return a list of query parameters that were removed, in the order
that they appear in the original URL (including duplicates).

Test: WTF_URL.URLRemoveQueryParameters

* Source/WTF/wtf/URL.cpp:
(WTF::removeQueryParameters):
* Source/WTF/wtf/URL.h:
* Tools/TestWebKitAPI/Tests/WTF/URL.cpp:
(TestWebKitAPI::TEST_F):

Canonical link: <a href="https://commits.webkit.org/258755@main">https://commits.webkit.org/258755@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9aead6d4fe3878745fd1ac182a69f0ccc3c93606

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102836 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11955 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35863 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/112091 "Build was cancelled. Recent messages:Failed to print configuration; Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Updated wpe dependencies; Compiled WebKit (cancelled)") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/172311 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/106799 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12972 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2864 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/95084 "Build was cancelled. Recent messages:Failed to print configuration; Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Updated gtk dependencies; Compiled WebKit (cancelled)") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109761 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108610 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/9969 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/93166 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37604 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91814 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24697 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/79338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/93044 "Found 1 new JSC stress test failure: wasm.yaml/wasm/lowExecutableMemory/exports-oom.js.default-wasm (failure)") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5408 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26112 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/89404 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/3094 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5556 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/2565 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29675 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11575 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45602 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/92327 "Built successfully") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6021 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7301 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20678 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Tools/TestWebKitAPI/Tests/WTF/URL.cpp:
```diff
@@ -540,39 +541,66 @@ TEST_F(WTF_URL, URLRemoveQueryParameters)
+        if (areEqual)
+            return;
+
+        WTFLogAlways("Test failed at %s:%d", __FILE__, lineNumber);
```
Did you mean to keep those?

whsieh replied:
Yes, that was my intent (as a debugging aid to help folks debug this test in the future, if needed).

cdumez commented on Source/WTF/wtf/URL.cpp:
```diff
@@ -1307,15 +1307,15 @@ bool isEqualIgnoringQueryAndFragments(const URL& a, const URL& b)
 }
 
-void removeQueryParameters(URL& url, const HashSet<String>& keysToRemove)
+Vector<String> removeQueryParameters(URL& url, const HashSet<String>& keysToRemove)
```
Hopefully this isn't used in performance sensitive code paths.

whsieh replied:
Hm...not as far as I know (AFAIK, in all the contexts where this helper method is used, appending some strings to a `Vector` should be very cheap, compared to the rest of the work those contexts). If this shows up in traces, we can surely add a different version of this that doesn't return anything, however.

cdumez commented on Source/WTF/wtf/URL.h:
```diff
@@ -259,7 +259,7 @@ WTF_EXPORT_PRIVATE bool protocolHostAndPortAreEqual(const URL&, const URL&);
 WTF_EXPORT_PRIVATE Vector<KeyValuePair<String, String>> queryParameters(const URL&);
 WTF_EXPORT_PRIVATE bool isEqualIgnoringQueryAndFragments(const URL&, const URL&);
-WTF_EXPORT_PRIVATE void removeQueryParameters(URL&, const HashSet<String>&);
+WTF_EXPORT_PRIVATE Vector<String> removeQueryParameters(URL&, const HashSet<String>&);
```
Maybe a comment to explain what this returns exactly.

The fact that it may return duplicates for e.g. should probably be documented too.

whsieh replied:
Good point — I'll add a comment explaining this:

```
// Returns the parameters that were removed (including duplicates), in the order that they appear in the URL.
```

whsieh committed:
Augment a WTF::URL helper function to return a list of query parameters
https://bugs.webkit.org/show_bug.cgi?id=250402

Reviewed by Chris Dumez.

Tweak `removeQueryParameters` to return a list of query parameters that were removed, in the order
that they appear in the original URL (including duplicates).

Test: WTF_URL.URLRemoveQueryParameters

* Source/WTF/wtf/URL.cpp:
(WTF::removeQueryParameters):
* Source/WTF/wtf/URL.h:
* Tools/TestWebKitAPI/Tests/WTF/URL.cpp:
(TestWebKitAPI::TEST_F):

Canonical link: https://commits.webkit.org/258755@main
